### PR TITLE
PACA-759: FISS service date query using from and to dates

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-759](https://jira.cms.gov/browse/PACA-759)

**User Story or Bug Summary:**

To have more claims available for end user's we should increase the fields in which we filter service-date on.

---

### What Does This PR Do?

`ResourceTypeV2` now uses a `List` of service date field names instead of a single name.  FISS claims use both from and to date fields while MCS claims only use to dates.

Note: Some of the changes in this PR are also present in PACA-719.  That PR is on hold while we evaluate a fix from the RDA API team that might make it unnecessary but those changes improved test coverage and readability so this PR builds on them rather than what is in master.

List of all changes:

- Removes duplicate method from QueryUtils and changes callers to use the one that's been retained.
- Moves logic to build general date range queries out of ClaimDao and into QueryUtils.
- Adds QueryUtilsTest class containing unit tests for the new method.
- Changes ResourceTypeV2 to have a list property for service date field names instead of a single field holding only one name.
- Changes ClaimDao logic to handle the case when there are multiple service date field names by combining the predicates for each date using logical OR.
- Changes FISS resource types to include both from and to dates instead of just to date.  (MCS still uses only to date)
- Adds a tiny fix to FHIR validation script to fix tests failing when only one resource has been modified.
- Modifies bfd-server-war pom to include lombok and jupiter dependencies.
- Modifies test cases to verify the new logic.
- Adds lombok to bfd-server-war to generate boilerplate code.
- Adds mockito jupiter package to bfd-server-war to simplify using mocks in unit tests.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/ClaimResponseTypeV2.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
           RdaMcsClaim.class,
           RdaMcsClaim.Fields.mbiRecord,
           RdaMcsClaim.Fields.idrClmHdIcn,
-          RdaMcsClaim.Fields.idrHdrToDateOfSvc,
+          List.of(RdaMcsClaim.Fields.idrHdrToDateOfSvc),
```
Comment (ericbuckley) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/ClaimResponseTypeV2.java:38]: This isn't in the JIRA issue, and I don't want to change this now.  But I'm thinking it makes sense to also include `idrHdrFromDateOfSvc` here.  Based on some early queries, I don't think it will make much of a difference, but from a readability perspective, I think it will make more sense to future maintainers.

**Review by ericbuckley — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/ClaimResponseTypeV2.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/ClaimResponseTypeV2.java:38]: Sounds good.  Now that the logic is in place it'd be trivial to add it any time.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java [hunk 2]
```diff
@@ -163,7 +235,6 @@ Predicate createMbiPredicate(
       Path<?> root,
       String mbiSearchValue,
       boolean isMbiSearchValueHashed,
-      boolean isOldMbiHashEnabled,
```
Comment (KyRobbins) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java:166]: Need to update javadoc

**Review by KyRobbins — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java:166]: Thanks.  I have fixed that and a few more.

**Review by ericbuckley — DISMISSED** [review]

**Review by KyRobbins — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDaoIT.java [hunk 3]
```diff
@@ -61,7 +72,7 @@ public void verifyQueryWithUnknownMbiFindsNothing() {
   public void verifyQueryWithKnownMbiHashFindsMatch() {
     ClaimDao claimDao = new ClaimDao(testUtils.getEntityManager(), metricRegistry, false);
     testUtils.seedData(false);
-    var claims = runFissMbiHashQuery(claimDao, RDATestUtils.MBI_HASH);
+    var claims = runMcsMbiHashQuery(claimDao, RDATestUtils.MBI_HASH);
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDaoIT.java:75]: can you help me understand the significance of these changes in this PR?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java [hunk 4]
```diff
@@ -133,6 +129,82 @@ public <T> List<T> findAllByMbiAttribute(
...
+      CriteriaBuilder builder,
+      Root<T> root,
+      DateRangeParam serviceDate,
+      List<String> serviceDateAttributeNames) {
+    final var serviceDatePredicates =
+        serviceDateAttributeNames.stream()
+            .map(root::<LocalDate>get)
+            .map(dateColumn -> serviceDatePredicate(builder, serviceDate, dateColumn))
+            .toArray(Predicate[]::new);
+    return builder.or(serviceDatePredicates);
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java:205]: The first map calls root.get(x) to get a Path for the date attribute and the serviceDatePredicate converts that into a Predicate.  The toArray just combines them all into an array of predicates.  Using multiple maps is generally clearer since each lambda is trivial rather than having a multi-line lambda that does both things.   The only other alternative would be a loop and I find those to be noisier and harder to follow than simple chained calls.

Are the method references confusing?  IDEA indicates they are preferred to simple lambdas so that's why I used them.  Personally I find simple lambdas clearer.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  /** Used when mocking query construction. */
+  @Mock private Predicate lowerBoundPredicate;
+  /** Used when mocking query construction. */
+  @Mock private Predicate upperBoundPredicate;
+  /** Used when mocking query construction. */
+  @Mock private Predicate notNullPredicate;
+
+  /**
+   * Test {@link QueryUtils#createDateRangePredicate} with no bounds in the {@link DateRangeParam}.
+   */
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java:38]: The test is 3 lines long.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  @Test
+  public void testCreateDateRangePredicateNoBounds() {
+    DateRangeParam dateRange = new DateRangeParam((DateParam) null, null);
+    QueryUtils.createDateRangePredicate(builder, dateRange, dateExpression);
+    verify(builder).and();
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java:43]: Given that it follows immediately from a call to `verify()` it seems pretty clear.  All of the tests follow this pattern.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  /**
+   * Test {@link QueryUtils#createDateRangePredicate} with both an upper and lower bound {@link
+   * DateRangeParam}.
+   */
+  @Test
+  public void testCreateDateRangePredicateBothBounds() {
+    DateRangeParam dateRange =
+        new DateRangeParam(
+            new DateParam(ParamPrefixEnum.GREATERTHAN, LOWER_DATE),
+            new DateParam(ParamPrefixEnum.LESSTHAN, UPPER_DATE));
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/QueryUtilsTest.java:119]: The comparison logic cases are tested separately.  This test just verifies that if multiple params are present they are both applied.  It doesn't need to fully reimplement the comparison logic.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java:205]: I think it's a mix between the method names, chained maps, and the method reference syntax. I would generally agree with IntelliJ that the method ref syntax is nicer, but when generics get involved and the object name is not super obvious what it is (like "root") it becomes a little more confusing. Also yeah at least to me .toArray(Predicate[]::new); is kinda hard to parse quickly.

Maybe just nix the shorthand on the first map and the toArray and it might read a little nicer without devolving it all into loops.

[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java:205]: I agree the JPA criteria query classes are pretty hard to follow.  Is there a naming convention on the team for how to name CriteriaQuery, Path, and Root objects?

[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/common/ClaimDao.java]: I have replaced the method refs with lambdas and given their parameters meaninful names.  Generally I'm not a fan of method refs anyway.  I've changed my IDEA setting to mark those opportunities as weak warnings instead of warnings to shut it up. :-)

**Review by christrotteradhoc — APPROVED** [review]

**Review by ericbuckley — APPROVED** [review]
